src/System.Threading.Tasks.Parallel/src/System/Threading/Tasks/ParallelLoopState.cs

```diff
@@ -0,0 +1,628 @@
+        /// </summary>
+        /// <param name="sharedParallelStateFlags">A flag shared among all threads participating
+        /// in the execution of a certain loop.</param>
+        internal ParallelLoopState64(ParallelLoopStateFlags64 sharedParallelStateFlags)
+            : base(sharedParallelStateFlags)
+        {
+            _sharedParallelStateFlags = sharedParallelStateFlags;
+        }
+
+        /// <summary>
+        /// Tracks the current loop iteration for the owning task.
+        /// This is used to compute whether or not the task should
+        /// terminate early due to a Break() call.
+        /// </summary>
+        internal long CurrentIteration
+        {
+            // No interlocks needed, because this value is only accessed in a single thread.
+            get { return _currentIteration; }
+            set { _currentIteration = value; }
+        }
+
+        /// <summary>
+        /// Returns true if we should be exiting from the current iteration
+        /// due to Stop(), Break() or exception.
+        /// </summary>
+        internal override bool InternalShouldExitCurrentIteration
+        {
+            get { return _sharedParallelStateFlags.ShouldExitLoop(CurrentIteration); }
+        }
+
+        /// <summary>
+        /// Returns the lowest iteration at which Break() has been called, or
+        /// null if Break() has not yet been called.
+        /// </summary>
+        internal override long? InternalLowestBreakIteration
+        {
+            // We don't need to worry about torn read/write here because
+            // ParallelStateFlags64.LowestBreakIteration property is protected
+            // by an Interlocked.Read().
+            get { return _sharedParallelStateFlags.NullableLowestBreakIteration; }
+        }
+
+        /// <summary>
+        /// Communicates that parallel tasks should stop when they reach a specified iteration element.
+        /// (which is CurrentIteration of the caller).
+        /// </summary>
+        /// <exception cref="T:System.InvalidOperationException">Break() called after Stop().</exception>
+        /// <remarks>
+        /// Atomically sets shared StoppedBroken flag to BROKEN, then atomically sets shared 
+        /// LowestBreakIteration to CurrentIteration, but only if CurrentIteration is less than 
+        /// LowestBreakIteration.
+        /// </remarks>
+        internal override void InternalBreak()
+        {
+            ParallelLoopState.Break(CurrentIteration, _sharedParallelStateFlags);
+        }
+    }
+
+    /// <summary>
+    /// State information that is common between ParallelStateFlags class
+    /// and ParallelStateFlags64 class.
+    /// </summary>
+    internal class ParallelLoopStateFlags
+    {
+        internal const int PLS_NONE = 0;
+        internal const int PLS_EXCEPTIONAL = 1;
+        internal const int PLS_BROKEN = 2;
+        internal const int PLS_STOPPED = 4;
+        internal const int PLS_CANCELED = 8;
```

Comment:
These constants aren't following our constant naming convention.  I think they'd need to be ParallelLoopStateNone, ParallelLoopStateExceptional, etc., or maybe PlsNone, PlsExceptional, etc.  They could even be moved into an enum.